compiler/rustc_mir_dataflow/src/impls/single_enum_variant.rs

```diff
@@ -0,0 +1,171 @@
+
+//use crate::lattice::FactArray;
+use crate::lattice::FactCache;
+use rustc_target::abi::VariantIdx;
+
+use rustc_middle::mir::*;
+
+use crate::{Analysis, AnalysisDomain};
+
+/// A dataflow analysis that tracks whether an enum can hold 0, 1, or more than one variants.
+pub struct SingleEnumVariant<'a, 'tcx> {
+    tcx: TyCtxt<'tcx>,
+    body: &'a mir::Body<'tcx>,
+}
+
+impl<'tcx> AnalysisDomain<'tcx> for SingleEnumVariant<'_, 'tcx> {
+    /// For each local, keep track of which enum index it is, if its uninhabited, or unknown.
+    //type Domain = FactArray<Fact, 128>;
+    type Domain = FactCache<Local, Location, VariantIdx, 16>;
+
+    const NAME: &'static str = "single_enum_variant";
+
+    fn bottom_value(&self, _: &mir::Body<'tcx>) -> Self::Domain {
+        //FactArray { arr: [Fact::TOP; 128] }
+        FactCache::new(Local::from_u32(0), Location::START, VariantIdx::MAX)
+    }
+
+    fn initialize_start_block(&self, body: &mir::Body<'tcx>, state: &mut Self::Domain) {
+        // assume everything is top initially.
+        let local_decls = body.local_decls();
+        for (l, _) in local_decls.iter_enumerated() {
+            state.remove(l);
+            //state.insert(l, Fact::TOP);
+            //state[l] = Fact::TOP;
+        }
+    }
+}
+
+impl<'tcx> SingleEnumVariant<'_, 'tcx> {
+    pub fn new<'a>(tcx: TyCtxt<'tcx>, body: &'a mir::Body<'tcx>) -> SingleEnumVariant<'a, 'tcx> {
+        SingleEnumVariant { tcx, body }
+    }
+    #[inline]
+    pub fn is_tracked(&self, place: &Place<'tcx>) -> bool {
+        place.ty(self.body, self.tcx).ty.is_enum()
+    }
+    fn assign(
+        &self,
+        state: &mut <Self as AnalysisDomain<'tcx>>::Domain,
+        lhs: &Place<'tcx>,
+        rhs: &Operand<'tcx>,
+        location: Location,
+    ) {
+        let _: Option<_> = try {
+            if !self.is_tracked(lhs) {
+                return;
+            }
+            let lhs_local = lhs.local_or_deref_local()?;
+
+            let new_fact = match rhs {
+                Operand::Copy(rhs) | Operand::Move(rhs) => {
+                    if let Some(rhs_local) = rhs.local_or_deref_local() {
+                        state.get(rhs_local).map(|f| f.1).copied()
+                    } else {
+                        rhs.ty(self.body, self.tcx).variant_index.map(|var_idx| var_idx)
+                    }
+                }
+                // Assigning a constant does not affect discriminant?
+                Operand::Constant(_c) => return,
+            };
+            if let Some(new_fact) = new_fact {
+                state.insert(lhs_local, location, new_fact);
+            } else {
+                state.remove(lhs_local);
+            }
+        };
+    }
+}
+
+impl<'tcx> Analysis<'tcx> for SingleEnumVariant<'_, 'tcx> {
+    fn apply_statement_effect(
+        &self,
+        state: &mut Self::Domain,
+        statement: &Statement<'tcx>,
+        loc: Location,
+    ) {
+        let (place, fact) = match &statement.kind {
```

Comment:
Please add a comment indicating what the contract associated with the `(place, fact)` is